hudi-client/src/main/java/org/apache/hudi/table/UpgradeDowngradeHelper.java

```diff
@@ -0,0 +1,202 @@
+          upgradeFromZeroToOne(config, jsc.hadoopConfiguration(), jsc);
+        } else if (versionFromPropertyFile == HoodieTableVersion.ZERO && toVersion == HoodieTableVersion.ONE) {
+          downgradeFromOneToZero();
+        } else {
+          throw new HoodieException("Illegal state wrt table versions. Version from proerpty file " + versionFromPropertyFile + " and current version " + toVersion);
+        }
+      }
+    }
+
+    /**
+     * If table version needs to be updated in hoodie.properties file.
+     * Step1: Copy hoodie.properties to hoodie.properties.orig
+     * Step2: add table.version to hoodie.properties
+     * Step3: delete hoodie.properties.orig
+     */
+    if (updateTableVersionInPropertyFile) {
+      updateTableVersionInHoodiePropertyFile(metaClient, toVersion);
+    }
+  }
+
+  /**
+   * Upgrade steps to be done to upgrade from hoodie table version 0 to 1.
+   */
+  private static void upgradeFromZeroToOne(HoodieWriteConfig config, Configuration hadoopConf, JavaSparkContext jsc) {
+    // fetch pending commit info
+    HoodieTable table = HoodieTable.create(config, hadoopConf);
+    HoodieTimeline inflightTimeline = table.getMetaClient().getCommitsTimeline().filterPendingExcludingCompaction();
+    List<String> commits = inflightTimeline.getReverseOrderedInstants().map(HoodieInstant::getTimestamp)
+        .collect(Collectors.toList());
+    for (String commit : commits) {
+      // for every pending commit, delete old marker files and re-SparkMaincreate marker files in new format
+      recreateMarkerFiles(commit, table, jsc);
+    }
+  }
+
+  public static void recreateMarkerFiles(final String commitInstantTime, HoodieTable table, JavaSparkContext jsc) throws HoodieRollbackException {
+    try {
+      Option<HoodieInstant> commitInstantOpt = Option.fromJavaOptional(table.getActiveTimeline().getCommitsTimeline().getInstants()
+          .filter(instant -> HoodieActiveTimeline.EQUALS.test(instant.getTimestamp(), commitInstantTime))
+          .findFirst());
+      if (commitInstantOpt.isPresent()) {
+        MarkerFiles markerFiles = new MarkerFiles(table, commitInstantTime);
+        markerFiles.quietDeleteMarkerDir();
+
+        List<HoodieRollbackStat> rollbackStats = new CopyOnWriteRollbackActionExecutor(jsc, table.getConfig(), table, "", commitInstantOpt.get(), false).mayBeRollbackAndGetStats(false);
+
+        for (HoodieRollbackStat rollbackStat : rollbackStats) {
+          for (FileStatus fileStatus : rollbackStat.getFilesToRollback()) {
+            String path = fileStatus.getPath().toString();
+            String dataFileName = path.substring(path.lastIndexOf("/") + 1);
+            markerFiles.create(rollbackStat.getPartitionPath(), dataFileName, path.endsWith(table.getBaseFileExtension()) ? IOType.CREATE : IOType.MERGE);
+          }
+          for (FileStatus fileStatus : rollbackStat.getCommandBlocksCount().keySet()) {
```

Comment:
I assume every entry in commandBlocks will be an APPEND. Correct me if I am wrong.